bot/Engine/BotController.py

```diff
@@ -0,0 +1,385 @@
+
+    def createOrderModel(self, symbol, order_params, order):
+        """ Create Order Model and fill only mandatory params. Other params are filled after order is filled. """
+        if 'price' not in order_params:
+            order_params['price'] = None
+        if 'take_profit_price' not in order_params:
+            order_params['take_profit_price'] = None
+        if 'stop_price' not in order_params:
+            order_params['stop_price'] = None
+        if 'is_entry' not in order_params:
+            order_params['is_entry'] = False
+        if order is None:
+            position_id = str(uuid4())
+        elif order is not None:
+            position_id = order.position_id
+
+        new_order_model = Order(
+                                id = str(uuid4()),
+                                position_id = position_id,
+                                bot_id = self.bot.id,
+                                symbol = symbol,
+                                price = order_params['price'],
+                                take_profit_price = order_params['take_profit_price'],
+                                order_type = order_params['order_type'],
+                                stop_price = order_params['stop_price'],
+                                original_quantity = order_params['quantity'],
+                                side = order_params['side'],
+                                is_entry =  order_params['is_entry'],
+                                is_test = self.test_mode,
+                                )
+        return new_order_model
+
+    def syncModels(self, pair, order, new_order_model):
+        """ Sync pairs and orders to new status """
+        if order is not None:        
+            new_order_model.matched_order_id = order.id
+            order.is_closed = True
+            order.matched_order_id = new_order_model.id
+            pair.active = False
+            pair.current_order_id = new_order_model.id
+
+        if order is None:
+            self.bot.current_balance = self.bot.current_balance - Decimal(self.bot.starting_balance) * Decimal(self.bot.entry_settings.initial_entry_allocation) / 100
+            pair.active = False
+            pair.current_order_id = new_order_model.id    
+
+    def computeQuantity(self, order):
+        """ compute exit quantity so that also partially filled orders can be handled."""
+        if order.side == 'BUY':
+            exit_quantity = order.executed_quantity
+        elif order.side == 'SELL':
+            exit_quantity = Decimal(order.original_quantity) - Decimal(order.executed_quantity)
+        return exit_quantity
+    
+    def simulateOrderInfo(self, order):
+        """ Used when BotController is in test mode. Simulates order info returned by exchange."""
+        order_status = dict()
+        candlestick_data = self.exchange.getSymbolKlines(order.symbol, self.kline_interval, self.strategy.minimum_period)
```

Comment:
Here my friend we need to look not only at the closing price, but also before that. Say the bot stays offline for the evening and we restart it in the morning. Well if an order was placed on the exchange, it could have been filled and we can ask the exchange for that info, but if it was a simulated one, we have no idea unless we look at the price movement between last night and today. To make this work, my idea would be, we can add another parameter in order called last_checked_time, and whenever we want to 'simulate' the market, we get all the data from last_checked_time until the present moment and look at it candle by candle.

Reply:
Yep completely agree!